src/libsyntax/parse/parser.rs

```diff
@@ -3432,7 +3432,16 @@ impl<'a> Parser<'a> {
         loop {
             pats.push(self.parse_pat()?);
             if self.check(&token::BinOp(token::Or)) { self.bump();}
-            else { return Ok(pats); }
+            else {
+                // Accidental use of || instead of | inbetween patterns
+                if self.token == token::OrOr {
+                    return Err(self.span_fatal_help(
```

Comment:
This is an instance of ["confusable tokens" pattern](https://github.com/rust-lang/rust/pull/46763#issuecomment-352178244), so we can easily do recovery instead of returning `Err`.

See https://github.com/rust-lang/rust/pull/46763 for the example of how to do it, except that we have `|` and `||` instead of `..` and `...` here.

Reply:
So I just need to emit the error instead?

Reply:
Yes, emit an error (non-fatal one) and continue parsing.

Reply:
Can you take a look now?